### PR TITLE
Xpu/weekly simple model enablement 2026 08 30

### PR DESCRIPTION
**Pull request by airMeng**
## Summary

Consolidates small XPU model-enablement and backend-parity changes from 11 source PRs onto the latest `main`.

Each source PR is kept as a separate commit for attribution and review. PR #33366 is intentionally excluded because its broad MiniMax-H3 diffusion changes are outside this simple-enablement scope.

## Consolidated PRs

- #34576 by @dayanandav - replace hardcoded CUDA model placement with the active device.
- #36803 by @cyxlily - enable the DeepSeek-V4 fused mHC post/pre gate on XPU when the optional `sgl_kernel` symbol is available.
~~- #36278 by @MeowMiaoJ - add Gemma3 RMSNorm XPU dispatch, tests, and benchmark coverage.~~
- #35244 by @pavansivaram - retain additional model-config shape regression coverage; the production fallback fix merged into `main` during this refresh.
~~- #33373 by @YangKai0616 - use the XPU MoE runner's expected MXFP4 expert scale dtype.~~
~~- #30218 by @jiayisunx - enable Qwen2-MoE shared-expert fusion on XPU.~~
- #29628 by @sureshnam - avoid duplicate Ministral3 construction during loading.
- #26592 by @jmunetong - add Gemma4 RMSNorm XPU coverage and CI registration.
- #36765 by @gaopengff - add the ERNIE mRoPE XPU Triton path.
~~- #30143 by @jiayisunx - enable Qwen3.5 alternate-stream execution on XPU.~~
~~- #33737 by @Amrutha-M05 - route MiniMax sparse decode top-k through the XPU kernel.~~

## Validation

- `pre-commit run --from-ref origin/main --to-ref HEAD`
- Model-config shape unit tests: 7 passed.
- Transformers fallback unit tests: 2 passed.
- Gemma4 RMSNorm XPU tests: 4 passed on `torch 2.12.1+xpu`.
- Parsed all changed Python files successfully.
- Verified DeepSeek-V4 safely disables fused mHC when the installed `sgl_kernel` lacks the optional XPU symbol.

The parameterized MiniMax sparse-decode XPU test did not complete after more than four minutes in local JIT compilation and was interrupted; CI should provide the authoritative result for that kernel path.

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :no_entry_sign: [Run #33584478534](https://github.com/sgl-project/sglang/actions/runs/33584478534)
Latest PR Test (Extra): :x: [Run #33584478468](https://github.com/sgl-project/sglang/actions/runs/33584478468)
Latest PR Test (AMD ROCm 7.2): :x: [Run #33584478526](https://github.com/sgl-project/sglang/actions/runs/33584478526)

### CODE REVIEW

**Comment by airMeng** [comment]
@mingfeima Here is the result of the first week

File: python/sglang/srt/layers/rotary_embedding/base.py [hunk 1]
```diff
@@ -454,8 +454,6 @@ def forward_xpu(
         ), "fused_set_kv_buffer_arg is not supported for xpu implementation"
         positions = torch.add(positions, offsets) if offsets is not None else positions
 
-        self._match_cos_sin_cache_dtype(query)
-
```
Comment (mingfeima) [inline python/sglang/srt/layers/rotary_embedding/base.py:458]: Cos/sin cache is often fp32 while query is bf16;

the CUDA path still casts. Restore unless the XPU fused kernel is proven to accept mixed dtypes.

**Review by mingfeima — CHANGES_REQUESTED** [review]

**Comment by mingfeima** [comment]
@airMeng still we have a few schedule things to settle:

1. how to handle requests for this combined PR? request original owner or let the agent handle it?
2. close the individual PRs once the combined PR is availble? after all, one of our goals is to reduce CI machine stress.

File: python/sglang/srt/layers/rotary_embedding/base.py (hunk 1 above)
Comment (gaopengff) [reply python/sglang/srt/layers/rotary_embedding/base.py:458]: Cuda uses fp32 cache by default. The kernel has supported mixed dtypes in xpu by https://github.com/sgl-project/sgl-kernel-xpu/pull/287.

**Comment by airMeng** [comment]
>how to handle requests for this combined PR? request original owner or let the agent handle it?

request original owners
>close the individual PRs once the combined PR is availble? after all, one of our goals is to reduce CI machine stress.

close the individual PRs once the combined PR merged. No need to trigger CI on the separate PRs if separate PRs have been included in the combine PR

File: test/registered/xpu/test_layernorm_xpu.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+import unittest
+
+import torch
+
+from sglang.test.ci.ci_register import register_xpu_ci
+
+register_xpu_ci(est_time=30, suite="stage-b-test-1-gpu-xpu")
+
+
+class TestGemma4RMSNormXPU(unittest.TestCase):
```
Comment (mingfeima) [inline test/registered/xpu/test_layernorm_xpu.py]: [nit] https://github.com/sgl-project/sglang/blob/main/docs/docs/developer_guide/contribution_guide.mdx#unit-tests-no-server-required

use `CustomTestCase` instead of raw unittest.TestCase.

File: python/sglang/srt/layers/rotary_embedding/base.py (hunk 1 above)
Comment (mingfeima) [reply python/sglang/srt/layers/rotary_embedding/base.py:458]: #287 covers `fused_qk_rope_with_cos_sin_cache_inplace`, so removing the cast is valid for the aligned fused branch.

The fallback calls the separate AOT `rotary_embedding` op, which still requires the cache dtype to match Q/K.

you can move it to fallback:
```python
if self.head_size in [128, 256, 512]:
    # FP32 cache supported by #287
    ...
else:
    self._match_cos_sin_cache_dtype(query)
    return torch.ops.sgl_kernel.rotary_embedding(...)
```

File: python/sglang/kernels/ops/attention/minimax_sparse/decode/flash_with_topk_idx.py [hunk 3]
```diff
@@ -1004,7 +1005,12 @@ def flash_decode_with_topk_idx(
         # Equivalent output to the 2-stage path (set of block ids, front-packed,
         # -1 padded); ~2-16x faster for long context. See
         # sglang/kernels/ops/attention/minimax_decode_topk.py.
-        from sglang.kernels.ops.attention.minimax_decode_topk import minimax_decode_topk
+        if is_xpu():
+            from sgl_kernel import minimax_decode_topk
+        else:
+            from sglang.kernels.ops.attention.minimax_decode_topk import (
+                minimax_decode_topk,
+            )
```
Comment (mingfeima) [inline python/sglang/kernels/ops/attention/minimax_sparse/decode/flash_with_topk_idx.py]: this depends on https://github.com/sgl-project/sgl-kernel-xpu/pull/371/changes which is still open. merge it first.

Comment (mingfeima) [reply python/sglang/kernels/ops/attention/minimax_sparse/decode/flash_with_topk_idx.py]: @airMeng

File: python/sglang/srt/models/ministral3.py [hunk 4]
```diff
@@ -141,6 +146,16 @@ def __init__(
...
 
+        # super().__init__ (LlamaModel) already built a full stack of LlamaDecoderLayers
+        # into self.layers; we rebuild them with Ministral3DecoderLayer below. Free the
+        # orphan stack BEFORE make_layers, otherwise both the old and the new stacks are
+        # resident while make_layers runs -> ~2x model weights -> OOM on memory-tight
+        # setups (e.g. a quantized 128B model on 4 GPUs). The modules linger via
+        # nn.Module ref-cycles, so a plain del isn't enough — collect explicitly.
+        del self.layers
+        gc.collect()
+        torch.get_device_module().empty_cache()
```
Comment (mingfeima) [inline python/sglang/srt/models/ministral3.py]: `torch.get_device_module()` returns `torch.cpu` on CPU, which has no `empty_cache()`.

Use `current_platform.empty_cache()` instead which has handles cpu platform.

File: python/sglang/srt/models/ernie45_moe_vl.py [hunk 5]
```diff
@@ -282,7 +282,7 @@ def forward(
         hidden_dim = hidden_states.shape[-1]
         hidden_states = hidden_states.view(-1, hidden_dim)
 
-        capturing = torch.cuda.is_current_stream_capturing()
+        capturing = torch.get_device_module().is_current_stream_capturing()
```
Comment (mingfeima) [inline python/sglang/srt/models/ernie45_moe_vl.py]: from sglang.srt.model_executor.runner import get_is_capture_mode

```
capturing = get_is_capture_mode()
```

**Review by mingfeima — CHANGES_REQUESTED** [review]

File: python/sglang/srt/layers/rotary_embedding/base.py (hunk 1 above)
Comment (gaopengff) [reply python/sglang/srt/layers/rotary_embedding/base.py:458]: Sure, I've moved the cast to fallback pass.

File: python/sglang/srt/models/ernie45_moe_vl.py (hunk 5 above)
Comment (gaopengff) [reply python/sglang/srt/models/ernie45_moe_vl.py]: Done.

Comment (airMeng) [reply python/sglang/srt/models/ernie45_moe_vl.py]: updated here

File: python/sglang/srt/layers/rotary_embedding/base.py (hunk 1 above)
Comment (airMeng) [reply python/sglang/srt/layers/rotary_embedding/base.py:458]: updated here

File: python/sglang/srt/models/ministral3.py (hunk 4 above)
Comment (mingfeima) [reply python/sglang/srt/models/ministral3.py]: @sureshnam

**Review by mingfeima — APPROVED** [review]
